gbasis/evals/density.py

```diff
@@ -95,12 +95,170 @@ def evaluate_density(one_density_matrix, basis, points, transform=None, threshol
+        and index 0 of the array for contractions.
+        Default is no transformation.
+
+
+    Returns
+    -------
+    dm_on_grid : np.ndarray(N1,N2,K_orb,K_orb)
+        Density Matrix evaluated at `N1,N2` grid points.
+
+    """
+
+    orb_evals = []
+    # evaluate basi(e)s on the point set(s)
+    for grid in points_list:
+        orb_eval = evaluate_basis(basis, grid, transform=transform)
+        orb_evals.append(orb_eval)
+    # if only one set of points is supplied, it is duplicated
+    if len(points_list) == 1:
+        orb_evals.append(orb_eval)
+
+    # Calulated performed using the evaluated orbitals
+    dm_on_grid = evaluate_dm_using_evaluated_orbs(one_density_matrix, orb_evals)
+
+    return dm_on_grid
+
+
+def evaluate_hole_x2(one_density_matrix, basis, points_list, transform=None):
+    r"""Return the two-body hole correlation function.
+
+    .. math ::
+
+        \left.
+        h(\mathbf{r}_1,\mathbf{r}_2) =
+        \right.
+        -\frac{|\gamma(\mathbf{r}_1,\mathbf{r}_2)|^2}
+        {\rho({\mathbf{r}_1})\rho({\mathbf{r}_2})}
+
+    Parameters
+    ----------
+    one_density_matrix : np.ndarray(K_orbs, K_orbs)
+        One-electron density matrix in terms of the given basis set.
+        If the basis is transformed using `transform` keyword, then the density matrix is assumed to
+        be expressed with respect to the transformed basis set.
+    basis : list/tuple of GeneralizedContractionShell
+        Shells of generalized contractions.
+    points_list : list of points [np.ndarray(N, 3)]
+        Cartesian coordinates of the points in space (in atomic units) where the basis functions
+        are evaluated.
+        Rows correspond to the points and columns correspond to the :math:`x, y, \text{and} z`
+        components.
+        This function can take a list of points at which basis functions are evaluated. If only one
+        set of points is given, it will be duplicated.
+    transform : np.ndarray(K_orbs, K_cont)
+        Transformation matrix from the basis set in the given coordinate system (e.g. AO) to linear
+        combinations of contractions (e.g. MO).
+        Transformation is applied to the left, i.e. the sum is over the index 1 of `transform`
+        and index 0 of the array for contractions.
+        Default is no transformation.
+
+
+    Returns
+    -------
+    hole_x2 : np.ndarray(N1,N2,K_orb,K_orb)
+        Two-body Exchange Hole evaluated at `N1,N2` grid points.
+
+    """
+    dens_evals = []
+
+    for grid in points_list:
+        dens_eval = evaluate_density(one_density_matrix, basis, grid, transform=transform)
+        dens_evals.append(dens_eval)
+
+    if len(points_list) == 1:
+        dens_evals.append(dens_eval)
+
+    # build density matrix on grid
+    dm_eval = evaluate_dm_density(one_density_matrix, basis, points_list, transform=transform)
+
+    # evaluate hole function
+    numerator = np.einsum("ijkl,ijkl->ijkl", dm_eval, dm_eval)
```

Comment:
If I am not wrong, you are multiplying dm_eval * dm_eval right? If this is the case, I was wondering if it would be more efficient to do `dm_eval * dm_eval`? I know using einsum is clear for notation so I am not against it.